data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferAccumulator.java

```diff
@@ -0,0 +1,181 @@
+import java.util.Optional;
+import org.mapdb.DB;
+import org.opensearch.dataprepper.plugins.s3keyindex.S3ObjectIndex;
+import org.opensearch.dataprepper.plugins.sink.S3SinkConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.waiters.S3Waiter;
+
+/**
+ * {@link BufferAccumulator} is responsible for upload accumulated data to
+ * amazon s3 bucket.
+ */
+public class BufferAccumulator {
+    
+    private static final Logger LOG = LoggerFactory.getLogger(BufferAccumulator.class);
+    private static final String LOCAL_FILE = "local_file";
+    private final S3Client s3Client;
+    private final S3SinkConfig s3SinkConfig;
+    private final NavigableSet<String> localEventSet;
+    private File fileAbsolutePath = null;
+    private String s3ObjectFileName = null;
+    private DB eventDataDb = null;
+    private StringBuilder eventBuilder;
+    private final String temporaryStorage;
+    private final int maxObjectUploadRetry;
+    private final String codecFileExtension;
+
+    /**
+     * @param localEventSet
+     * @param s3Client
+     * @param s3SinkConfig
+     * @param eventDataDb
+     * @param codecFileExtension
+     */
+    public BufferAccumulator(final NavigableSet<String> localEventSet, final S3Client s3Client,
+            final S3SinkConfig s3SinkConfig, DB eventDataDb, final String codecFileExtension) {
+        this.localEventSet = localEventSet;
+        this.s3Client = s3Client;
+        this.s3SinkConfig = s3SinkConfig;
+        this.eventDataDb = eventDataDb;
+        this.codecFileExtension = codecFileExtension;
+        temporaryStorage = s3SinkConfig.getTemporaryStorage();
+        maxObjectUploadRetry = s3SinkConfig.getMaxUploadRetries();
+    }
+
+    /**
+     * Writing data to local-file/in-memory based on temporary_storage configuration
+     */
+    public void doAccumulate() {
+        s3ObjectFileName = S3ObjectIndex.getIndexAliasWithDate(s3SinkConfig.getObjectOptions().getFilePattern()) + "."
+                + codecFileExtension;
+        eventBuilder = new StringBuilder();
+        if (temporaryStorage.equals(LOCAL_FILE)) {
+            localFileProcessor();
+        } else {
+            inMemoryProcessor();
+        }
+        if (eventDataDb != null && !eventDataDb.isClosed()) {
+            eventDataDb.close();
+        }
+    }
+
+    /**
+     * Store event data into local file
+     */
+    private void localFileProcessor() {

```

Comment:
Using the strategy pattern to decide how to handle local file versus in-memory can help make the code cleaner. You can have independent tests for each strategy.